### ISSUE TITLE
[Announcement] AllowSynchronousIO disabled in all servers

### ISSUE THREAD

**Issue by Tratcher** [dotnet/aspnetcore#7644]
AllowSynchronousIO is a option in each server that enables or disables sync IO APIs like HttpReqeuest.Body.Read, HttpResponse.Body.Write, Stream.Flush, etc.. These APIs have long been a source of thread starvation and application hangs. Starting in 3.0.0-preview3 these are disabled by default.

Affected servers:
- Kestrel
- HttpSys
- IIS in-process
- TestServer

Expect errors similar to:
- `Synchronous operations are disallowed. Call ReadAsync or set AllowSynchronousIO to true instead.`
- `Synchronous operations are disallowed. Call WriteAsync or set AllowSynchronousIO to true instead.`
- `Synchronous operations are disallowed. Call FlushAsync or set AllowSynchronousIO to true instead.`

Each server has a AllowSynchronousIO option that controls this behavior and the default for all of them is now false.

The behavior can also be overridden on a per request basis as a temporary mitigation. 
```c#
var syncIOFeature = HttpContext.Features.Get<IHttpBodyControlFeature>();
if (syncIOFeature != null)
{
    syncIOFeature.AllowSynchronousIO = true;
}
```

If you have trouble with TextWriters or other streams calling sync APIs in Dispose, call the new DisposeAsync API instead.

**Comment by kieronlanning** [comment]
As-per https://github.com/aspnet/Docs/issues/10983, will this be back ported to 2.x?

Currently setting this to false in 2.2 breaks response compression.

**Comment by davidfowl** [comment]
No, it won't be back ported. It's a massive breaking change.

**Comment by Tratcher** [comment]
Note response compression was fixed in 3.0 using the new DisposeAsync API.

**Comment by GusBeare** [comment]
I have been working on porting an angularJS app to .Net Core and just upgraded to Preview 3. After upgrade some of the client GET requests now blow with this error:

InvalidOperationException: Synchronous operations are disallowed. Call WriteAsync or set AllowSynchronousIO to true instead.

I don't get it because by default the angular requests are async and why do just some of them break?

I am using VS 2019 Preview and will be hosting IIS in process.

Any ideas?

**Comment by Tratcher** [comment]
@GusBeare what's the full stack trace?

**Comment by GusBeare** [comment]
I get the following html in the response.

[error.txt](https://github.com/aspnet/AspNetCore/files/2943373/error.txt)

**Comment by Tratcher** [comment]
@GusBeare let's take this to https://github.com/aspnet/AspNetCore/issues/8302

**Comment by Kaelum** [comment]
This change breaks the usage of `XmlSerializer.Deserialize(Stream)`.  Do we now need to change `AllowSynchronousIO` to `true`?  Or should we be performing serialization differently?  As stated by others, this is a significant break change over all previous versions of ASP.NET.

**Comment by Tratcher** [comment]
@Kaelum you should asynchronously pre-buffer the request body in this case. That's how MVC handles it:
https://github.com/aspnet/AspNetCore/blob/093df67c06297c20edb422fe6d3a555008e152a9/src/Mvc/Mvc.Formatters.Xml/src/XmlSerializerInputFormatter.cs#L102-L117

**Comment by Kaelum** [comment]
@Tratcher the `FileBufferingReadStream` constructor above doesn't appear to be in the 3.0 Preview 5 libraries.  Also, I have concerns that using `FileBufferingReadStream` might be to slow, if it is actually caching to the filesystem.  We need to process a minimum of 12,000 requests/sec, which we do by keeping everything in memory.  I tried using the `BufferedReadStream` object, but under the covers it is doing everything synchronously as well.  This is what I have tried:

Works, but I have concerns with caching to the file system, and the constructor in your example doesn't exist:
``` C#
using (FileBufferingReadStream readstream = new FileBufferingReadStream(request.Body, memoryThreshold, null, @"D:\Temp\", ArrayPool<byte>.Shared))
{
	await readstream.DrainAsync(CancellationToken.None);
	readstream.Seek(0L, SeekOrigin.Begin);

	XmlSerializer xmlSerializer = new XmlSerializer(typeof(XmlRequestBcap));
	xmlRequestBcap = xmlSerializer.Deserialize(readstream) as XmlRequestBcap;
}
```

Works, but probably non-performant:
``` C#
using (StreamReader streamReader = new StreamReader(request.Body, Encoding.UTF8))
using (StringReader stringReader = new StringReader(await streamReader.ReadToEndAsync()))
{
	XmlSerializer xmlSerializer = new XmlSerializer(typeof(XmlRequestBcap));
	xmlRequestBcap = xmlSerializer.Deserialize(stringReader) as XmlRequestBcap;
}
```

Does not work, executes synchronously:
``` C#
using (BufferedReadStream readstream = new BufferedReadStream(request.Body, memoryThreshold))
{
	if (!await readstream.EnsureBufferedAsync(CancellationToken.None))
	{
		throw new InvalidOperationException("Request body is empty.");
	}

	XmlSerializer xmlSerializer = new XmlSerializer(typeof(XmlRequestBcap));
	xmlRequestBcap = xmlSerializer.Deserialize(readstream) as XmlRequestBcap;
}
```

Variation of above that works, adding a MemoryStream, but is it safe and performant:
``` C#
using (BufferedReadStream readstream = new BufferedReadStream(request.Body, memoryThreshold))
{
	if (!await readstream.EnsureBufferedAsync(CancellationToken.None))
	{
		throw new InvalidOperationException("Request body is empty.");
	}

	using (MemoryStream memoryStream = new MemoryStream(readstream.BufferedData.Array, readstream.BufferedData.Offset, readstream.BufferedData.Count, false))
	{
		XmlSerializer xmlSerializer = new XmlSerializer(typeof(XmlRequestBcap));
		xmlRequestBcap = xmlSerializer.Deserialize(memoryStream) as XmlRequestBcap;
	}
}
```

If `FileBufferingReadStream` is the most performant, and recommended for our case, what happened to the constructor that I see in your source reference?  It is definitely not in the published 3.0 preview.

Thanks!

**Comment by davidfowl** [comment]
> @Tratcher the FileBufferingReadStream constructor above doesn't appear to be in the 3.0 Preview 5 libraries. Also, I have concerns that using FileBufferingReadStream might be too slow, if it is actually caching to the filesystem. We need to process a minimum of 12,000 requests/sec, which we do by keeping everything in memory. I tried using the BufferedReadStream object, but under the covers it is doing everything synchronously as well. This is what I have tried:

It's there AFAIK.

**Comment by Kaelum** [comment]
@davidfowl & @Tratcher I just verified that it is not on the `SDK 3.0.100-preview5-011568` build.  The referenced **Microsoft.AspNetCore.WebUtilities** assembly is pointing to:

C:\Program Files\dotnet\packs\Microsoft.AspNetCore.App.Ref\3.0.0-preview5-19227-01\ref\netcoreapp3.0\Microsoft.AspNetCore.WebUtilities.dll

When I disassemble that assembly, I can very definitely see only 4 constructors, with the 2 parameter constructor missing.

**Comment by Tratcher** [comment]
Here's the MVC PR: https://github.com/aspnet/AspNetCore/pull/9806
I think that was a preview6 change.

**Comment by FreyJim** [comment]
@Tratcher  
I have the same issue by using streams with moq framework. I changed already inside the moq setup to call the async methods and also added the section to the startup class. 
`services.Configure<TestServer>(options =>
{
        options.AllowSynchronousIO = 
});` 

But I still get the same error message: 
System.InvalidOperationException: Synchronous operations are disallowed. Call WriteAsync or set AllowSynchronousIO to true.

Do I have to do something special by using the `TestServer`?

**Comment by Tratcher** [comment]
What's the stack trace?

**Comment by Tratcher** [comment]
How are you creating TestServer in your test? The old pattern was to do `new TestServer(webhostbuilder);` which wouldn't use DI to create an instance so `services.Configure<TestServer>(options => { options.AllowSynchronousIO = });` wouldn't apply. If you are newing it up directly then you can directly set the option as well.

The new pattern looks like this and would work with DI options:
https://github.com/aspnet/AspNetCore/blob/8c02467b4a218df3b1b0a69bceb50f5b64f482b1/src/Hosting/TestHost/test/TestServerTests.cs#L53-L62

**Comment by FreyJim** [comment]
I'm doing it like this. 
The TestServer is from `microsoft.aspnetcore.testhost\3.0.0-preview7.19365.7`
`
var webHostBuilder = WebHost.CreateDefaultBuilder().UseStartup<Startup>();

            using (var testServer = new TestServer(webHostBuilder))
            using (var httpClient = testServer.CreateClient())
            {
                var response = await httpClient.GetAsync("api/cloudconnector/logs").ConfigureAwait(false);
                response.EnsureSuccessStatusCode();
`

the `Startup` is like the start of the web api itself. 

maybe it could be related to the mock which is used inside the test. 
`
mock.Setup(x => x.WriteZipLogToStream(It.IsAny<Stream>()))
                    .Callback<Stream>(async s =>
                    {
                        using (var writer = new StreamWriter(s))
                        {
                            await writer.WriteAsync("1234567");
                        }
                    });
'

**Comment by YetaWF** [comment]
I understand the intention. But this was a seriously bad idea. It breaks pretty much all use of 3rd party libraries in my code. Now I have to go an retest everything and add enabling code. Thanks a bunch.

**Comment by manigandham** [comment]
@YetaWF You can just enable the setting again if you need it.

**Comment by YetaWF** [comment]
Sure thing. Already did that. After spending too much time researching this. This was not mentioned on  https://docs.microsoft.com/en-us/aspnet/core/migration/22-to-30

We're not all asp.net core fan bois and spend our waking moments following the development. We have real jobs where we have to solve business problems. So something like this just wastes our time. 

Of course the solution is always easy, once you know the solution. And since you know how the "enable the setting" again, it would have been nice to show us HOW (for us mere mortals who are not as enlightened). As in:

public void ConfigureServices(IServiceCollection services) {
   // If using Kestrel:
   services.Configure<KestrelServerOptions>(options =>
   {
      options.AllowSynchronousIO = true;
   });
   // If using IIS:
   services.Configure<IISServerOptions>(options =>
   {
      options.AllowSynchronousIO = true;
   });
   ....

**Comment by manigandham** [comment]
The breaking change is listed under Kestrel:
https://docs.microsoft.com/en-us/aspnet/core/migration/22-to-30#allowsynchronousio-disabled

**Comment by YetaWF** [comment]
I stand corrected. But like I said, not a fan boi, so I skimmed over that not seeing the implications. I have real work to do...

**Comment by manigandham** [comment]
@YetaWF Reading the docs is expected when moving up to a new major release with breaking changes, and you can always ask for help, but please stop with the silly attitude. We all have "real work" to do. Disparaging the team and other users trying to help isn't nice or productive.

**Comment by YetaWF** [comment]
BTW love that "Breaking changes will happen in a new release". I have Windows API code written in 1994 that still works on Windows 10 (that's 25 years ago). So it doesn't have to be that way apparently.

**Comment by MaddoxDevelopment** [comment]
You don't have to be a fan boy to know how to check for breaking changes on a major release. That's called being a decent developer.

**Comment by davidfowl** [comment]
This is a good change because it'll make developers think twice before doing blocking IO which can cause thread pool starvation in many applications. This is just the first step, expect to see more in this area (analyzers etc). There should be action taken on either the library side or the app side to opt into this bad behavior. Before it was subtle and would result in unreliable applications, now it'll throw and you have to turn on the "I don't care" flag to work around it. Now that flag can be used to audit suspicious applications doing sync IO.

**Comment by YetaWF** [comment]
Indeed, it's a good feature. I actually love that it does this. BUT it kind of flew by under my radar, which caused my frustration.
And yes, if you're developing new code, this is a great way to insure you're not doing something stupid. But if you have a substantial code base that you just want to upgrade from 2.x to 3.0, this will most certainly bite you in the a**. 
If you're an overworked (& underpaid) developer, you'll definitely skim over the Kestrel async mention in the migration doc (the Borg are about perfection, I'm about getting things done), and then you'll find out that things don't work.

Now understand that I'm very mindful of sync/async I/O, BUT I use third party stuff I simply have no control over. And that's what failed. For example, System.ServiceModel.Syndication Rss20FeedFormatter WriteTo caused an error. It's synchronous. (I think that's some MS code).
Also some code that writes ZIP files had issues, Both instances, rarely used features in my sites.
That's when I realized what was going on and of course found the local and global "switches" for AllowSynchronousIO.
I also ran across some issue (there was apparently even one in an asp.net core 3 preview) where the amount of data being written was causing an issue. So even at 100% code coverage testing, you still might not see the issue until you have large data I/O.

But why would I EVER want AllowSynchronousIO disabled in PRODUCTION??? That is the default now. That means if I do something stupid (sync I/O) in my code now I crash 100% of the time. Prior to 3.0 I crashed only under high server load (thread starvation). Improvement? I think not.

Yes it's a great feature to audit my code. But I don't want to AUDIT my code in PRODUCTION.
And introducing a "breaking" change shouldn't be about "Hey, your code will now crash for sure". Breaking changes used to be about "your code won't compile any more". Changes that I painstakingly have to verify by 100% code coverage and edge-case testing are not simply "breaking" changes. They're "fu**-you" changes. LOL. (Apologies, I couldn't find a better description).

Obviously now that I know all this, it's a simple matter to adjust my code (audit in debug, don't audit in production). But everyone with an existing non-trivial codebase upgrading to 3.0 will have to go through this "learning" process. Let's say 100,000 developers @ 2 hours each. aspnet core 3.0 has just wasted a shi*-ton of development time. (not quite a crap-ton, but there's always 4.0).

I would love to see more features like this. They're really helpful in finding issues with my code.
But please don't force them down my throat in PRODUCTION. Better yet, let me log such errors in production, to my log file. (THIS is really something I could use). But don't force me to crash, or force me to find all these issues in development.  

Unfortunately, the way it's set up now, most people will simply turn AllowSynchronousIO off once they hit the first issue, and that defeats the purpose.

**Comment by davidfowl** [comment]
I get it, we cause pain and I'm sorry. The workaround is quite simple and I hope you didn't lose too much sleep finding it😄. Hopefully, we can do a bit more static analysis next time with analyzers to catch a few more things at build time. 

I also get that it's hard to test every code path before deploying to production but this *was* a major version upgrade after a year of development with some significant features added. 

That said, I'm sorry again it caused you pain. As you find these issues in other libraries, let us know and let library authors know. They are bugs that need to be fixed and this change will hopefully cause the masses to influence blocking code 😄 (at least that's my hope).

> Unfortunately, the way it's set up now, most people will simply turn AllowSynchronousIO off once they hit the first issue, and that defeats the purpose.

When they hit the issue, it's the first thing I'll ask 😉.

**Comment by YetaWF** [comment]
The workaround is actually counter productive. It just hides the problem. I don't think that's helpful.
Letting me log these errors in production, to my log file, would we 100% more helpful. The way it is now, I have no choice but to enable sync in production (I'll never now what problems I actually hit) and disable sync in development (and hope to hit the problems that may exist). We all know I won't hit 100% of the problems in development. Reality says so.

But keep up the good work, keep the analyzers and code checks coming. I'll make them work somehow. Spelling out the actual implications of some new defaults explicitly will help too.

**TL/DR**: But if at all possible, give me some interface so I can LOG stuff that is detrimental to my app, like sync I/O, rather than making the app crash. We're all on the same side. We want to make web apps that work.

**Comment by davidfowl** [comment]
Not really there are other options like buffering (which we ourselves had to implement as a result of this feature for things like JSON.NET). 

That’s good feedback.

**Comment by jhudsoncedaron** [comment]
Got here, same error message. We built our application server to stream out of blocking objects because it is much faster than running up and down the stack many times per request. Setting AllowSynchronousIO does work, but has to be set for both IIS and Kestrel because the hosting platform can change.

**Comment by davidfowl** [comment]
Why don’t you buffer the request then?

**Comment by jhudsoncedaron** [comment]
@davidfowl : Because this way the caller can begin to consume it before it's even finished coming out of the DB.

**Comment by davidfowl** [comment]
> @davidfowl : Because this way the caller can begin to consume it before it's even finished coming out of the DB.

Ah so you prefer to block a thread pool thread than buffer? How big is the payload?

**Comment by jhudsoncedaron** [comment]
@davidfowl : A few MB. I prefer to spend the resource I have (a MB of RAM per request for the blocked thread) over a resource I don't have (another CPU core).

**Comment by davidfowl** [comment]
@jhudsoncedaron Just be aware that i'll affect your scale if you have too many of those operations happening at the same time. It might seem like a good tradeoff until all of your threads are blocked and you can't do any further operations.

**Comment by MaddoxDevelopment** [comment]
This also errors if you are using Newtonsoft Json. I attempted to make the switch to the new json api's, but there was too much of a change needed for it to work properly, so I'll have to stick with syncronous IO enabled for now.

**Comment by davidfowl** [comment]
@MaddoxDevelopment This is what our JSON.NET InputFormatter does https://github.com/aspnet/AspNetCore/blob/387bd1a4047b6f98b50a6a9d58878febac0b6f34/src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs#L134-L147

**Comment by MaddoxDevelopment** [comment]
> @MaddoxDevelopment This is what our JSON.NET InputFormatter does
> 
> https://github.com/aspnet/AspNetCore/blob/387bd1a4047b6f98b50a6a9d58878febac0b6f34/src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs#L134-L147

Interesting. I was getting the synchronous IO error right after I added 

```
services.AddMvc()
    .AddNewtonsoftJson();
```
It seems that it is working fine for most of my endpoints except for this particular one.

https://d.pr/i/u1oirh

The user variable is just a regular C# object, so not sure why it would be throwing.

**Comment by jhudsoncedaron** [comment]
@davidfowl : The task scheduler fails catastrophically when the computer runs out of memory, and there's no way around it. So we had to buy memory instead of CPU. We're currently running with the task scheduler allowed to allocate unlimited threads to avoid thread starvation.

**Comment by davidfowl** [comment]
@MaddoxDevelopment if you have a repro then send me the details with steps, that shouldn't happen.

@jhudsoncedaron Most things fail when the computer runs out of memory and cpu.

**Comment by jhudsoncedaron** [comment]
@davidfowl : I'm just accustomed to `throw OutOfMemoryException()` rather than deadlock in native code.

**Comment by pranavkm** [comment]
@MaddoxDevelopment do you use `request.EnableBuffering()` in your app?

**Comment by benmccallum** [comment]
How would this look in the context of a `WebApplicationFactory` test? 

I'm doing the following in Startup under `ConfigureServices`, but I'm guessing the server is different in this scenario?

```
services.Configure<KestrelServerOptions>(options =>
{
    options.AllowSynchronousIO = true;
});
```

**Comment by Tratcher** [comment]
@benmccallum `WebApplicationFactory.Server.AllowSynchronousIO = true;`
https://github.com/aspnet/AspNetCore/blob/a4d9faefe2ed6fbebd500413032336607ec92778/src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs#L73
https://github.com/aspnet/AspNetCore/blob/a4d9faefe2ed6fbebd500413032336607ec92778/src/Hosting/TestHost/src/TestServer.cs#L90

**Comment by benmccallum** [comment]
Sigh, how did I not see that. Was about to browse the source. Cheers again, sorry for the spam!

**Comment by Muchiachio** [comment]
How to know which code part is throwing exception if we get this stack trace?
We don't really use our own text writers, only those passed to `IHtmlContent.WriteTo` method.
```
System.InvalidOperationException: Synchronous operations are disallowed. Call WriteAsync or set AllowSynchronousIO to true instead.
   at Microsoft.AspNetCore.Server.IIS.Core.HttpResponseStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at Microsoft.AspNetCore.Server.IIS.Core.WrappingStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at Microsoft.AspNetCore.WebUtilities.HttpResponseStreamWriter.FlushInternal(Boolean flushEncoder)
   at Microsoft.AspNetCore.WebUtilities.HttpResponseStreamWriter.Dispose(Boolean disposing)
   at System.IO.TextWriter.Dispose()
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ActionContext actionContext, IView view, ViewDataDictionary viewData, ITempDataDictionary tempData, String contentType, Nullable`1 statusCode)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewResultExecutor.ExecuteAsync(ActionContext context, ViewResult result)
   at Microsoft.AspNetCore.Mvc.ViewResult.ExecuteResultAsync(ActionContext context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|29_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeResultFilters()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

**Comment by Tratcher** [comment]
```
   at Microsoft.AspNetCore.WebUtilities.HttpResponseStreamWriter.FlushInternal(Boolean flushEncoder)
   at Microsoft.AspNetCore.WebUtilities.HttpResponseStreamWriter.Dispose(Boolean disposing)
   at System.IO.TextWriter.Dispose()
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ActionContext actionContext, IView view, ViewDataDictionary viewData, ITempDataDictionary tempData, String contentType, Nullable`1 statusCode)
```
@NTaylorMullen ?

**Comment by Muchiachio** [comment]
I pulled all the sources and found the problem. A 3rd party library had a bug which threw an exception in a partial view rendered by `Html.Partial` (no async to sync code used). It's still strange that it was caught as a sync IO problem only when exception was thrown. I can share a repro in a separate issue if @NTaylorMullen is interested.

**Comment by Tratcher** [comment]
@Muchiachio please open a new issue for this scenario, it sounds like we need to fix something in MVC.

**Comment by NTaylorMullen** [comment]
> I pulled all the sources and found the problem. A 3rd party library had a bug which threw an exception in a partial view rendered by `Html.Partial` (no async to sync code used). It's still strange that it was caught as a sync IO problem only when exception was thrown. I can share a repro in a separate issue if @NTaylorMullen is interested.

@Muchiachio Filing an issue with a repro would be a great start!

**Comment by RomanTymchyshyn** [comment]
> @MaddoxDevelopment if you have a repro then send me the details with steps, that shouldn't happen.

@MaddoxDevelopment , @davidfowl , so did you figured out what the problem was in that case?

I have a similar problem, I use `.AddNewtonsoftJson`, but then receiving "Synchronous operations are disallowed. Call ReadAsync or set AllowSynchronousIO to true instead.".
You can see in the stack trace, that `NewtonsoftJsonInputFormatter` gets called.

StackTrace:
`at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.Read(Byte[] buffer, Int32 offset, Int32 count)
       at Microsoft.AspNetCore.WebUtilities.FileBufferingReadStream.Read(Byte[] buffer, Int32 offset, Int32 count)
       at Microsoft.AspNetCore.WebUtilities.HttpRequestStreamReader.ReadIntoBuffer()
       at Microsoft.AspNetCore.WebUtilities.HttpRequestStreamReader.Read(Char[] buffer, Int32 index, Int32 count)
       at Newtonsoft.Json.JsonTextReader.ReadData(Boolean append, Int32 charsRequired)
       at Newtonsoft.Json.JsonTextReader.ParseValue()
       at Newtonsoft.Json.JsonReader.ReadAndMoveToContent()
       at Newtonsoft.Json.JsonReader.ReadForType(JsonContract contract, Boolean hasConverter)
       at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
       --- End of stack trace from previous location where exception was thrown ---
       
       at Microsoft.AspNetCore.Mvc.Formatters.NewtonsoftJsonInputFormatter.ReadRequestBodyAsync(InputFormatterContext context, Encoding encoding)
       at Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinder.BindModelAsync(ModelBindingContext bindingContext)
       at Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.BindModelAsync(ActionContext actionContext, IModelBinder modelBinder, IValueProvider valueProvider, ParameterDescriptor parameter, ModelMetadata metadata, Object value)
       at Microsoft.AspNetCore.Mvc.Controllers.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d.MoveNext()
       --- End of stack trace from previous location where exception was thrown ---
       
       at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
       at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)`

**Comment by olaj** [comment]
@benmccallum @Tratcher About the TestServer implementation. I have tried to set that AllowSynchronousIO property on various places but i only get a Stack Overflow exception in the program.cs, or i get the error "Synchronous operations are disallowed. Call ReadAsync or set AllowSynchronousIO to true instead.". Is there any example somewhere on how to actually set this property? Where do i set it?

My setup is similar to what they do in this tutorial. https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-3.0

**Comment by davidfowl** [comment]
Stackoverflow? Can you provide a repo?

**Comment by Tratcher** [comment]
@olaj did you try the sample I gave above https://github.com/aspnet/AspNetCore/issues/7644#issuecomment-543833374?

**Comment by danstur** [comment]
@pranavkm We are hitting this issue with `EnableBuffering` active. A simplified scenario:
```
// Inside Middleware
var buffer = new byte[4000];
context.Request.EnableBuffering();
context.Request.Body.Position = 0;
var readBytes = await context.Request.Body.ReadAsync(buffer, 0, buffer.Length);
context.Request.Body.Position = 0;
```

results sometimes, but non-deterministically (afaics) in an InvalidOperationException with `Synchronous IO APIs are disabled, see AllowSynchronousIO` and the following stack trace:
```
at int Microsoft.AspNetCore.Server.HttpSys.RequestStream.Read(out byte[] buffer, int offset, int size)
   at int Microsoft.AspNetCore.WebUtilities.FileBufferingReadStream.Read(byte[] buffer, int offset, int count)
   at int Microsoft.AspNetCore.WebUtilities.HttpRequestStreamReader.ReadIntoBuffer()
   at int Microsoft.AspNetCore.WebUtilities.HttpRequestStreamReader.Read(char[] buffer, int index, int count)
   at int Newtonsoft.Json.JsonTextReader.ReadData(bool append, int charsRequired)
   at bool Newtonsoft.Json.JsonTextReader.ParseObject()
   at void Newtonsoft.Json.JsonReader.Skip()
   at void Newtonsoft.Json.Serialization.JsonSerializerInternalReader.HandleError(JsonReader reader, bool readPastError, int initialDepth)
   at object Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateObject(object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, string id)
   at object Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, object existingValue)
   at object Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, object existingValue)
   at object Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, bool checkAdditionalContent)
   at async Task<InputFormatterResult> Microsoft.AspNetCore.Mvc.Formatters.NewtonsoftJsonInputFormatter.ReadRequestBodyAsync(InputFormatterContext context, Encoding encoding)
   at async Task Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinder.BindModelAsync(ModelBindingContext bindingContext)
   at async Task<ModelBindingResult> Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.BindModelAsync(ActionContext actionContext, IModelBinder modelBinder, IValueProvider valueProvider, ParameterDescriptor parameter, ModelMetadata metadata, object value)
   at void Microsoft.AspNetCore.Mvc.Controllers.ControllerBinderDelegateProvider+<>c__DisplayClass0_0+<<CreateBinderDelegate>g__Bind|0>d.MoveNext()
   at async Task Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()+Awaited(?)
   at async Task Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeNextExceptionFilterAsync()+Awaited(?)
```
are we using the API incorrectly or should we create a new issue for this?

**Comment by Tratcher** [comment]
You're only doing a single ReadAsync on the body, you need to read until the end of the body to ensure it's fully buffered. Read in a loop until it returns 0.

**Comment by pha3z** [comment]
If I understand this change correctly, I appreciate it.  But I am not clear on how to pre-buffer the request body to access it in other parts of asp.net.

I implemented my own HMAC Authentication Handler that I add with services.AddAuthentication().  I want to confirm that the body has not been tampered with in transit, so I encrypt Request.Body along with other headers on both the client and the server.  The problem is that I can't read Request.Body in the Authentication Handler method since reading the stream directly can interfere with the rest of the pipeline.

It would be nice if there was a simple switch that you can use in Startup to automatically buffer every request before sending anything down the pipeline.  That way Authentication Handlers or whatever else you want to use have access to the entire Request Body before the controller method is ever executed.  Is this possible?

The only examples I find when I search for pre-buffering Request.Body talk about doing it inside the controller method. Clearly, that's not helpful.

**Comment by Tratcher** [comment]
HttpRequest.EnableBuffering is the simplest mechanism.
https://github.com/dotnet/aspnetcore/blob/c39ecb9f9b387ec39112c3eace5ae38db06df160/src/Http/Http/src/Extensions/HttpRequestRewindExtensions.cs#L21
The example [above](https://github.com/dotnet/aspnetcore/issues/7644#issuecomment-657216298) mostly covered it, except that it needs to read to the end of the body.

> It would be nice if there was a simple switch that you can use in Startup to automatically buffer every request before sending anything down the pipeline. That way Authentication Handlers or whatever else you want to use have access to the entire Request Body before handlers are ever executed. Is this possible?

Yes, via middleware using the example above.

Note we don't recommend this or make it any easier because it has side-effects like no longer being able to override request body size limits per endpoint.

**Comment by davidfowl** [comment]
@Tratcher lets make sure to get this documented.

**Comment by Tratcher** [comment]
Filed https://github.com/dotnet/AspNetCore.Docs/issues/20001

**Comment by matthiaslischka** [comment]
Is there a way to enable it for specific calls? We do use a third party component that does not support async calls and it feels wrong to set the whole application to `AllowSynchronousIO` just for this one call.

**Comment by Tratcher** [comment]
@matthiaslischka Yes:

> The behavior can also be overridden on a per request basis as a temporary mitigation.
> 
> ```cs
> var syncIOFeature = HttpContext.Features.Get<IHttpBodyControlFeature>();
> if (syncIOFeature != null)
> {
>     syncIOFeature.AllowSynchronousIO = true;
> }
> ```

**Comment by matthiaslischka** [comment]
We have this scenario where we create a zip file for some very small files and stream it directly to the HttpContext response stream.
The library we use does not support asynchronous calls.

Without drastically changing the implementation we now can 
1. allow synchronous IO operations for this specific call or
2. use a memory stream and then at the end copy the memory stream to the HttpContext response stream

I'm not sure which one is better/cleaner.
Not sure if (2) is a good solution or just a dirty workaround. Does it improve the IO load?

**Comment by Tratcher** [comment]
@matthiaslischka If the files are small then buffering to a memory stream is better. It avoids accidentally blocking threads with IO, and it will probably perform better because it will aggregate small writes from the zip library.
